### ISSUE TITLE
Getting 404 Not Found errors

### ISSUE THREAD

**Issue by dawbarton** [skilion/onedrive#299]
I've been getting a lot of 404 Not Found errors during syncing which seem to stop the sync from completing. At first I though it was because I was modifying (deleting) files simultaneously on two computers but I've found that I'm also getting it when I don't modify anything on my laptop but just use onedrive to download updates.

The typical output I get looks like
```
Moving Teaching/emat2-stats/exam/emat2-stats-2018-ans.tex to Teaching/emat2-stats/exam/emat2-stats-2018-ans.tex~
Downloading Teaching/emat2-stats/exam/emat2-stats-2018-ans.tex~... done.
Uploading Teaching/emat2-stats/exam/emat2-stats-2018-ans.tex~...HTTP request returned status code 404 (Not Found)
{
    "error": {
        "code": "itemNotFound",
        "innerError": {
            "date": "2018-02-20T17:31:58",
            "request-id": "272b39b6-4549-42ac-9b1f-60eb7460af4e"
        },
        "message": "The resource could not be found."
    }
}
Uploading Teaching/emat2-stats/exam/emat2-stats-2018-ans.tex~...HTTP request returned status code 404 (Not Found)
{
    "error": {
        "code": "itemNotFound",
        "innerError": {
            "date": "2018-02-20T17:32:19",
            "request-id": "f3768fff-8ef7-4c77-8476-8701be42a322"
        },
        "message": "The resource could not be found."
    }
}
Uploading Teaching/emat2-stats/exam/emat2-stats-2018-ans.tex~...onedrive.OneDriveException@src/onedrive.d(408): HTTP request returned status code 404 (Not Found)
{
    "error": {
        "code": "itemNotFound",
        "innerError": {
            "date": "2018-02-20T17:32:37",
            "request-id": "bf25c89c-1994-42b6-8911-fdfa21dda261"
        },
        "message": "The resource could not be found."
    }
}
----------------
src/onedrive.d:33 void onedrive.OneDriveApi.checkHttpCode(ref const(std.json.JSONValue)) [0x572458]
src/onedrive.d:352 std.json.JSONValue onedrive.OneDriveApi.upload(immutable(char)[], immutable(char)[]) [0x571e11]
src/onedrive.d:156 std.json.JSONValue onedrive.OneDriveApi.simpleUploadReplace(immutable(char)[], immutable(char)[], immutable(char)[], const(char)[]) [0x570dd6]
src/sync.d:537 void sync.SyncEngine.uploadFileDifferences(itemdb.Item, immutable(char)[]) [0x57b13c]
src/sync.d:468 void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x57aab5]
src/sync.d:488 void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0x57ac6d]
src/sync.d:465 void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x57aa84]
src/sync.d:488 void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0x57ac6d]
src/sync.d:465 void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x57aa84]
src/sync.d:488 void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0x57ac6d]
src/sync.d:465 void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x57aa84]
src/sync.d:488 void sync.SyncEngine.uploadDirDifferences(itemdb.Item, immutable(char)[]) [0x57ac6d]
src/sync.d:465 void sync.SyncEngine.uploadDifferences(itemdb.Item) [0x57aa84]
src/sync.d:440 void sync.SyncEngine.scanForDifferences(immutable(char)[]) [0x57a92b]
src/main.d:195 void main.performSync(sync.SyncEngine) [0x56975b]
src/main.d:121 _Dmain [0x56928b]
```

Any suggestions? I'm happy to look through the source (I'm on v1.1.1) if someone can give me an idea where to start.

**Comment by dawbarton** [comment]
I think this problem is to do with files that are deleted and then recreated with the same name but a different unique identifier. This causes confusion with the onedrive client with it downloading the updated file, then deleting it(!) and then registering a conflict.

**Comment by abraunegg** [comment]
There is a bunch of updates in #314 which may assist here. Can you retry your testing & using #314 as your base?

**Comment by dawbarton** [comment]
Thanks for this but unfortunately when I start from a clean install (initial sync with OneDrive) I get
```
db9052@tweety onedrive $ ./onedrive --synchronize -v
Loading config ...
Using Config Dir: /home/db9052/.config/onedrive
Initializing the OneDrive API ...
Authorize this app visiting:<removed for brevity>
Enter the response uri: <removed>
Opening the item database ...
All operations will be performed in: /home/db9052/OneDrive-emm
Initializing the Synchronization Engine ...
Account Type: business
Default Drive ID: b!nE5hxLnrmUCR_Z9bG3G0sNE1KWw4NpNJpIvHradqSobDcwLd5IegSaQ_Y7Rg5vgk
Default Root ID: 01FTEU43V6Y2GOVW7725BZO354PWSELRRZ
Remaining Free Space: 1099507518913
Applying changes of Path ID: 01FTEU43V6Y2GOVW7725BZO354PWSELRRZ
core.exception.AssertError@src/itemdb.d(293): Assertion failure
----------------
??:? _d_assertp [0x59ef25]
src/itemdb.d:293 immutable(char)[] itemdb.ItemDatabase.computePath(const(char)[], const(char)[]) [0x57663d]
src/sync.d:364 void sync.SyncEngine.applyDifference(std.json.JSONValue, immutable(char)[], bool) [0x58fbc6]
src/sync.d:303 void sync.SyncEngine.applyDifferences(immutable(char)[], const(char)[]) [0x58f864]
src/sync.d:167 void sync.SyncEngine.applyDifferences() [0x58f0d5]
src/main.d:313 void main.performSync(sync.SyncEngine, immutable(char)[], bool, bool, bool) [0x579722]
src/main.d:206 _Dmain [0x57910c]
```
When I use the original client (without the PR) it works fine (on the initial sync at least).

**Comment by dawbarton** [comment]
If I use the original client (latest tagged version) to do the initial sync and then use the PR version to do further syncs, this seems to work in the basic use-case and doesn't give any 404 errors.

However, the underlying problem still appears to be there. Steps to reproduce are

1. Create test.txt with something in it (doesn't matter what). 
2. Sync.
3. *Via the web sharepoint interface* (or another computer) move test.txt to test-old.txt and create a new file called test.txt with something different in it.
4. Sync.

After the sync in 4. I'm left with test-old.txt, test-<hostname>.txt, and test-<hostname>-2.txt indicating that something went very wrong with conflict resolution (there shouldn't be any conflicts since nothing changed locally between 2. and 4.

Any help gratefully received!

**Comment by abraunegg** [comment]
Thanks for the steps to reproduce. Let me look into it further tomorrow.

The issue will be that the file with the same name exists in the DB - but the meta data & everything else is different based on what OneDrive is returning.

**Comment by abraunegg** [comment]
@dawbarton

I followed your exact reproduction steps and cannot reproduce this issue. The steps I performed were
1. Create a new file locally (test.txt)
2. Sync (in this case a business account)
```
onedrive --synchronize --verbose
Loading config ...
Using Config Dir: /root/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /root/OneDrive
Initializing the Synchronization Engine ...
Account Type: business
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 1099509948715
Applying changes of Path ID: <redacted>
Fetching details for OneDrive Root
OneDrive Root exists in the database
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Uploading file ./test.txt... done.
Uploading file ./test.txt... done.
Remaining free space: 1099509948698
Applying changes of Path ID: <redacted>
```
3. _Via O365 / OneDrive Business web interface_ rename test.txt to test-old.txt. Create a new file called test.txt, new file contents, different from original test.txt
4. Sync 
```
onedrive --synchronize --verbose
Loading config ...
Using Config Dir: /root/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /root/OneDrive
Initializing the Synchronization Engine ...
Account Type: business
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 1099509931702
Applying changes of Path ID: <redacted>
The local item has a different modified time 2018-Apr-19 23:10:06Z remote is 2018-Apr-19 23:11:59Z
The local item has a different hash
The local item is out of sync, renaming...
Downloading test.txt...OneDrive Response: '200 - OK'
 done.
Downloading test.txt... done.
The local item has a different modified time 2018-Apr-19 23:11:59Z remote is 2018-Apr-19 23:10:06Z
The local item has a different hash
The local item is unsynced, renaming
Downloading test-old.txt...OneDrive Response: '200 - OK'
 done.
Downloading test-old.txt... done.
Fetching details for OneDrive Root
OneDrive Root exists in the database
Uploading differences of .
Processing root
The directory has not changed
Processing test.txt
The file has been deleted
Deleting item from OneDrive: test.txt
Processing test-old.txt
The file has not changed
Uploading new items of .
Uploading file ./test-[hostname]-2.txt... done.
Uploading file ./test-[hostname]-2.txt... done.
Remaining free space: 1099509931685
Uploading file ./test-[hostname].txt... done.
Uploading file ./test-[hostname].txt... done.
Remaining free space: 1099509931668
Applying changes of Path ID: <redacted>
```

I am not getting the db assertion that you experienced.

I am seeing the 3 file differences however. What I would expect to happen is this on the sync after the web UI changes:

Get item metadata & compare 'test.txt'
Local version 'test.txt' moved to test-[hostname].txt (local name conflict)
items.db updated to reflect the local file name change
Download 'test.txt'
Get item metadata & compare 'test-old.txt'
Metadata matches updated items.db data for test-[hostname].txt
Local version 'test-[hostname].txt' moved to test-old.txt (local name resolution)
items.db updated to reflect the local file name change
Skip download as 'test-old.txt' matches

This means we should end up with the same 2 files on either side

**Comment by dawbarton** [comment]
The db assertion happens when I use your modified client to do the initial set-up. I.e., delete everything (items database, refresh token, files, etc) and then set it up from scratch. This is separate from my sync problem.

I've just started from a clean system, pulled your git repository, built the client, connected to my OneDrive for Business account and then I get the db assertion.

I'll check the sync problem again in a moment.

**Comment by abraunegg** [comment]
Hi David,

I literally just posted fix for the db assertion - i was keeping my items.sqlite3 & not deleting it - which is why I didnt pick up the assertion. Commit 66d00c5ec9e00b57e1e237477e9e12c80d46763c fixes the assertion issue.

Best regards,

Alex

**Comment by dawbarton** [comment]
Great, thanks for your efforts on this!

As for the sync issue; it's still there for me.
1. Create test.txt
2. Sync
```
Loading config ...
Using Config Dir: /home/db9052/.config/onedrive
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/db9052/OneDrive-emm
Initializing the Synchronization Engine ...
Account Type: business
Default Drive ID: [removed]
Default Root ID: [removed]
Remaining Free Space: 1099507357179
Applying changes of Path ID: [removed]
Fetching details for OneDrive Root
OneDrive Root exists in the database
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Uploading file ./test.txt... done.
Uploading file ./test.txt... done.
Remaining free space: 1099507357167
Applying changes of Path ID: [removed]
```
3. Move test.txt to test-old.txt via web.
4. Sync (no local changes made)
```
Loading config ...
Using Config Dir: /home/db9052/.config/onedrive
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/db9052/OneDrive-emm
Initializing the Synchronization Engine ...
Account Type: business
Default Drive ID: [removed]
Default Root ID: [removed]
Remaining Free Space: 1099507302975
Applying changes of Path ID: [removed]
The local item has a different modified time 2018-Apr-20 08:41:31Z remote is 2018-Apr-20 08:44:05Z
The local item has a different hash
The local item is out of sync, renaming...
Downloading test.txt...OneDrive Response: '200 - OK'
 done.
Downloading test.txt... done.
The local item has a different modified time 2018-Apr-20 08:44:05Z remote is 2018-Apr-20 08:41:31Z
The local item has a different hash
The local item is unsynced, renaming
Downloading test-old.txt...OneDrive Response: '200 - OK'
 done.
Downloading test-old.txt... done.
Fetching details for OneDrive Root
OneDrive Root exists in the database
Uploading differences of .
Processing root
The directory has not changed
Uploading new items of .
Uploading file ./test-tweety.cityinthesky.co.uk.txt... done.
Uploading file ./test-tweety.cityinthesky.co.uk.txt... done.
Remaining free space: 1099507302963
Uploading file ./test-tweety.cityinthesky.co.uk-2.txt... done.
Uploading file ./test-tweety.cityinthesky.co.uk-2.txt... done.
Remaining free space: 1099507250392
Applying changes of Path ID: [removed]
```
I'm left with
```
 test-old.txt
 test-tweety.cityinthesky.co.uk-2.txt
 test-tweety.cityinthesky.co.uk.txt
```
when I should just have test.txt and test-old.txt.

**Comment by abraunegg** [comment]
Yep .. that is the next round of fixes I will have to look at - correct conflict resolution

**Comment by dawbarton** [comment]
OK, great. What is confusing me is that there isn't really a conflict in the sense that nothing changes locally between the syncs...

**Comment by sbocquet** [comment]
From what we have just seen with @abraunegg, there may be a sync problem from OneDrive to local since :
https://github.com/abraunegg/onedrive/commit/77c0cdbe24b1360b82685b4e5842df2ae23e7999

Last one that fully works for me is : https://github.com/abraunegg/onedrive/commit/ec3573ea1d5f8addcbd1937be7b58f8254495e02

**Comment by abraunegg** [comment]
There was a bad commit which was resolved with reverts & new patches over the weekend. The latest version is dff245d29bf09c979d5bc40e82b852b48c957ac5 which fixes the introduced regression.

Will have to work on the conflict naming aspect next.

If the original issue (getting 404 errors) is resolved in the latest #314 commit can you close this case out?

For the conflict resolution item - if you can open a new case for tracking that would be great.
